### PR TITLE
[PSM Interop] Fail test if client or server pods restarted during test

### PR DESCRIPTION
**Pull request by XuanWang-Amos**
Fail test if client or server pods restarted during test.

#### Testing
Tested locally, test will fail with message similar to:
```
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/xuanwn/workspace/xds/grpc/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py", line 501, in tearDown
    ))
AssertionError: 5 != 0 : Server pods unexpectedly restarted {sever_restarts} times during test.

----------------------------------------------------------------------
Ran 1 test in 886.867s
```

### CODE REVIEW

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py [hunk 1]
```diff
@@ -573,3 +573,16 @@ def _replicas_available(cls, deployment: V1Deployment, count: int) -> bool:
         return (isinstance(deployment, V1Deployment) and
                 deployment.status.available_replicas is not None and
                 deployment.status.available_replicas >= count)
+
+
+def get_pod_restarts(k8s_namespace: KubernetesNamespace,
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:578]: Let's make it a part of the class, same as other methods. I know it's not perfect, but we should refactor the whole thing anyway. At least we'll be consistent in being imperfect.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 2]
```diff
@@ -494,6 +495,26 @@ def cleanup(self):
         self.server_runner.cleanup(force=self.force_cleanup,
                                    force_namespace=self.force_cleanup)
 
+    def _check_pod_restart_time(self) -> None:
+        # We should fail test if pod restarted during test (b/269192257).
+        logger.info('Checking Pods restart times')
+        client_restarts = k8s.get_pod_restarts(self.client_runner.k8s_namespace,
+                                               self.client_runner.deployment)
+        self.assertEqual(
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:503]: We should not assert here, just get the data - so the cleanup can happen. After the cleanup, we should do the assertions, in a `finally` blocks.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 3]
```diff
@@ -480,6 +480,7 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
 
     def tearDown(self):
         logger.info('----- TestMethod %s teardown -----', self.id())
+        self._check_pod_restart_time()
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:483]: Todo: the same check in `XdsUrlMapTestCase`.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py (hunk 1 above)
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:578]: Originally I did put it into the class, but got some complains from sanity tests, let me try again.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 4]
```diff
@@ -480,20 +482,37 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
 
     def tearDown(self):
         logger.info('----- TestMethod %s teardown -----', self.id())
+        # Todo(xuanwn): Add same check to XdsUrlMapTestCase.
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:485]: Note: I recently made a [PR](https://github.com/grpc/grpc/pull/33090) that adds retry logic to the  `XdsUrlMapTestCase` similar to what's happening here. Don't forget to sync when my PR is merged before adding this check to the URL Map.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_client_runner.py [hunk 5]
```diff
@@ -165,6 +165,17 @@ def _xds_test_client_for_pod(self, pod: k8s.V1Pod, *,
...
                              rpc_host=rpc_host)
 
+    def get_pod_restarts(self) -> int:
+        total_restart: int = 0
+        pods: List[k8s.V1Pod]
+        if not self.k8s_namespace or not self.deployment:
+            return total_restart
+        pods = self.k8s_namespace.list_deployment_pods(self.deployment)
+        for pod in pods:
+            total_restart = sum(status.restart_count
```
[the remark this answers is not in the document]
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_client_runner.py:175]: Oh right, my mistake.

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_server_runner.py [hunk 6]
```diff
@@ -228,6 +228,17 @@ def _xds_test_server_for_pod(self,
                              secure_mode=secure_mode,
                              rpc_host=rpc_host)
 
+    def get_pod_restarts(self) -> int:
```
[the remark this answers is not in the document]
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_server_runner.py:231]: It's the same with client runner, but `KubernetesBaseRunner` don't have `deployment`, I'll move it to base runner and make `deployment` as a parameter then.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 4 above)
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:483]: Got it :)

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_server_runner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_xds_server_runner.py]: Great idea!

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 4 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:483]: This was merged BTW

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 7]
```diff
@@ -480,13 +480,29 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
+            self.server_runner.deployment)
         retryer = retryers.constant_retryer(wait_fixed=_timedelta(seconds=10),
                                             attempts=3,
                                             log_level=logging.INFO)
         try:
             retryer(self.cleanup)
         except retryers.RetryError:
             logger.exception('Got error during teardown')
+        finally:
+            # We should fail test if pod restarted during test (b/269192257).
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:497]: No need to link the bug here.
Let me offer a different description.

```suggestion
            # Fail if any of the pods restarted.
            # In most cases a pod restart indicates the test app has crashed.
```

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 8]
```diff
@@ -480,13 +480,29 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
+            self.assertEqual(client_restarts,
+                             0,
+                             msg=('Client pods unexpectedly restarted'
+                                  f'{client_restarts} times during test.'))
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:501]:
```suggestion
                                  f' {client_restarts} times during test.'))
```

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 9]
```diff
@@ -480,13 +480,29 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
+            self.assertEqual(server_restarts,
+                             0,
+                             msg=('Server pods unexpectedly restarted'
+                                  f'{server_restarts} times during test.'))
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:505]:
```suggestion
                                  f' {server_restarts} times during test.'))
```

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 4 above)
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:483]: Rebased and added test to `xds_url_map_testcase`.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 10]
```diff
@@ -480,13 +480,29 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
                                             log_level=logging.INFO)
         try:
             retryer(self.cleanup)
         except retryers.RetryError:
             logger.exception('Got error during teardown')
+        finally:
+            # We should fail test if pod restarted during test (b/269192257).
+            self.assertEqual(client_restarts,
+                             0,
+                             msg=('Client pods unexpectedly restarted'
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:500]: Thanks for the detailed message here! Very useful for debugging.
Let's also add a note suggesting what a pod restart means:

`In most cases, this is caused by the test client app crash.`

and 

`In most cases, this is caused by the test server app crash.`

**Review by sergiitk — COMMENTED** [review]
Getting close

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py [hunk 11]
```diff
@@ -213,6 +213,7 @@ class _MetaXdsUrlMapTestCase(type):
     # and tear down GCP resources.
     _started_test_cases = set()
     _finished_test_cases = set()
+    _pod_restart_time: int = 0
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:216]: no need for this, remember?

Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:216]: Sorry forgot to remove it.

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_base_runner.py [hunk 12]
```diff
@@ -98,6 +98,17 @@ def stop_pod_dependencies(self, *, log_drain_sec: int = 0):
 
         self.pod_log_collectors = []
 
+    def get_pod_restarts(self, deployment: k8s.V1Deployment) -> int:
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_base_runner.py:101]: If you want - I'm ok with overriding this at the client/server level, and just calling super().get_pod_restarts(self.deployment)

Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/runners/k8s/k8s_base_runner.py:101]: IMO it's fine to keep it since it's not that complicate.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 13]
```diff
@@ -480,13 +480,29 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
 
     def tearDown(self):
         logger.info('----- TestMethod %s teardown -----', self.id())
+        # Todo(xuanwn): Add same check to XdsUrlMapTestCase.
+        logger.debug('Getting pods restart times')
+        client_restarts: int = self.client_runner.get_pod_restarts(
+            self.client_runner.deployment)
+        server_restarts: int = self.server_runner.get_pod_restarts(
+            self.server_runner.deployment)
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:488]: Just realized this could raise an error too. Let's wrap it into a try and catch `except (retryers.RetryError, k8s.NotFound) as e:`

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py (hunk 11 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:216]: That's what PR reviews are for :)

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py (hunk 13 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:488]: And just log the error.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py [hunk 14]
```diff
@@ -378,6 +387,16 @@ def cleanupAfterTests(cls):
...
             logging.exception('Got error during teardown')
+        finally:
+            # Fail if any of the pods restarted.
+            cls.assertEqual(
+                client_restarts,
+                0,
+                msg=(
+                    'Client pods unexpectedly restarted'
+                    f' {client_restarts} times during test. '
+                    'In most cases, this is caused by the test client app crash.'
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:398]: Nit, same as above.
```suggestion
                    f' {client_restarts} times during test.'
                    ' In most cases, this is caused by the test client app crash.'
```

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 15]
```diff
@@ -480,13 +480,42 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
             logger.exception('Got error during teardown')
+        finally:
+            # Fail if any of the pods restarted.
+            self.assertEqual(
+                client_restarts,
+                0,
+                msg=(
+                    'Client pods unexpectedly restarted'
+                    f' {client_restarts} times during test. '
+                    'In most cases, this is caused by the test client app crash.'
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:509]: Latest internal formatters do that. I know I was doing the opposite before, but I really like this new way. It makes a natural indent - which makes it much clearer this is a continuation of the previous line. 

```suggestion
                    f' {client_restarts} times during test.'
                    ' In most cases, this is caused by the test client app crash.'
```

File: tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py [hunk 16]
```diff
@@ -480,13 +480,42 @@ def initKubernetesClientRunner(self) -> KubernetesClientRunner:
...
+                ))
+            self.assertEqual(
+                server_restarts,
+                0,
+                msg=(
+                    'Server pods unexpectedly restarted'
+                    f' {server_restarts} times during test. '
+                    'In most cases, this is caused by the test client app crash.'
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py:517]: Nit, same as above.

```suggestion
                    f' {server_restarts} times during test.'
                    ' In most cases, this is caused by the test client app crash.'
```

**Review by sergiitk — APPROVED** [review]
LGTM! Thanks for the change! I'll run the adhoc tests, then we can merge it.

**Comment by sergiitk** [comment]
Adhoc tests: 
- [ ] [grpc/core/v1.54.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/e5c32f7f-9742-408f-99d2-1fa36b86d071)
- [ ] [grpc/core/v1.54.x/branch/linux/psm-security](https://source.cloud.google.com/results/invocations/c06f0500-1c66-409f-9678-42475017a6f1)
- [ ] [grpc/core/v1.54.x/branch/linux/grpc_xds_url_map](https://source.cloud.google.com/results/invocations/696d8a3e-4805-414c-bdcc-f54d35f6a293)

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py [hunk 17]
```diff
@@ -378,6 +387,16 @@ def cleanupAfterTests(cls):
...
+        finally:
+            # Fail if any of the pods restarted.
+            cls.assertEqual(
+                client_restarts,
+                0,
+                msg=
+                ('Client pods unexpectedly restarted'
+                 f' {client_restarts} times during test.'
+                 ' In most cases, this is caused by the test client app crash.'
+                ))
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:399]: This failed with
```py
Traceback (most recent call last):
  File "/tmpfs/tmp/tmp.QBtlnvekrC/grpc/tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py", line 392, in cleanupAfterTests
    cls.assertEqual(
TypeError: assertEqual() missing 1 required positional argument: 'second'
```

I think we can't call `assertEqual` here, because it's a class :( 
I guess just do an `assert`.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py (hunk 17 above)
Comment (XuanWang-Amos) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py]: Changed to `assert`, new build:
- [ ] [grpc/core/v1.54.x/branch/linux/grpc_xds_url_map](https://source.cloud.google.com/results/invocations/1a202bc4-7d5d-45a4-b17d-8548f98ff745/targets/grpc%2Fcore%2Fv1.54.x%2Fbranch%2Flinux%2Fgrpc_xds_url_map)

Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py]: Pass

**Review by sergiitk — APPROVED** [review]
